src/main/java/com/android/volley/Request.java

```diff
@@ -320,17 +323,34 @@ public String getCacheKey() {
     }
 
     /**
-     * Mark this request as canceled.  No callback will be delivered.
+     * Mark this request as canceled.
+     *
+     * <p>No callback will be delivered as long as either:
+     * <ul>
+     *     <li>This method is called on the same thread as the {@link ResponseDelivery} is running
+     *     on. By default, this is the main thread.
+     *     <li>The request subclass being used overrides cancel() and ensures that it does not
+     *     invoke the listener in {@link #deliverResponse} after cancel() has been called in a
+     *     thread-safe manner.
+     * </ul>
+     *
+     * <p>There are no guarantees if both of these conditions aren't met.
      */
+    // @CallSuper
     public void cancel() {
-        mCanceled = true;
+        synchronized (mLock) {
+            mCanceled = true;
+            mErrorListener = null;
+        }
     }
 
     /**
      * Returns true if this request has been canceled.
      */
     public boolean isCanceled() {
-        return mCanceled;
+        synchronized (mLock) {
```

Comment:
I think we should have a separate lock object for this. I'm just worried that one thread is trying to cancel while another is handling the waiting requests from the RequestCompleteListener. Then the cancelling thread (which may be trying to cancel a whole bunch of requests) has to wait.

Reply:
What if we altered the notify methods to only grab the current NetworkRequestCompleteListener while holding the lock, while moving the dispatch itself out of the lock? Or is there a reason that the dispatch itself needs to be part of the same critical section?

Reply:
That would be better actually, since the RequestCompleteListener doesn't actually need the request, just the response, so there is no reason for the request to wait.